### PR TITLE
#80 add context_legacy.go and context.go for go 1.7 migration

### PR DESCRIPTION
**Pull request by shawnps**
Tests on Go 1.7 currently failing with:

```
➜  sessions [go17-context] go test
--- FAIL: TestFlashes (0.00s)
    sessions_test.go:71: No cookies. Header: map[]
FAIL
exit status 1
FAIL    github.com/gorilla/sessions 0.014s
```

Trying to debug. cc @elithrar

### CODE REVIEW

**Comment by elithrar** [comment]
I suspect this is because the copied `http.Request` isn't being passed back to the caller, and thus the context isn't retained, but need to dig into where that's being triggered (late here, so will look tomorrow).

**Comment by shawnps** [comment]
@elithrar thanks! I appreciate you taking the time to look. Let me know what I can do, if you don't have time to debug I will continue to try. Figured I'd just ping you since you probably have better insight on it than me.

**Comment by elithrar** [comment]
Still looking at this. Trying to debug where we forget to return any copy of the `http.Request` after doing a `contextSave`.

**Comment by shawnps** [comment]
@elithrar after returning the request in the `Get` methods, the tests pass for me locally (will see about travis in a bit). This of course changes the contract for the `Get` method, so I'm not sure how you'd like to handle that.

**Comment by elithrar** [comment]
I think that was inevitable as we have to return any copied Request.

Let me review tonight - thanks for working on this!

**Comment by elithrar** [comment]
Apparently I already came to these conclusions a while ago: https://github.com/gorilla/sessions/issues/80 

I'd like to tag a release prior to merging this, and then a release after, so anyone who is vendoring can vendor pre- / post- breaking change.

**Comment by shawnps** [comment]
Thanks, let me know if there's anything else I can do to help

**Comment by elithrar** [comment]
Tagged v1.1 before I break this. Will do a v1.2 on the weekend when I merge this.

File: store.go [hunk 1]
```diff
@@ -76,8 +76,12 @@ type CookieStore struct {
 //
 // It returns a new session and an error if the session exists but could
 // not be decoded.
-func (s *CookieStore) Get(r *http.Request, name string) (*Session, error) {
-	return GetRegistry(r).Get(s, name)
+func (s *CookieStore) Get(r *http.Request, name string) (*Session, *http.Request, error) {
```
Comment (elithrar) [inline store.go]: We shouldn't need to return the store here, if we're already replacing the `*http.Request` passed in.

File: store.go [hunk 2]
```diff
@@ -179,8 +183,12 @@ func (s *FilesystemStore) MaxLength(l int) {
 // Get returns a session for the given name after adding it to the registry.
 //
 // See CookieStore.Get().
-func (s *FilesystemStore) Get(r *http.Request, name string) (*Session, error) {
-	return GetRegistry(r).Get(s, name)
+func (s *FilesystemStore) Get(r *http.Request, name string) (*Session, *http.Request, error) {
```
Comment (elithrar) [inline store.go]: We shouldn't need to return the store here, if we're already replacing the `*http.Request` passed in.

File: store.go [hunk 3]
```diff
@@ -21,7 +21,7 @@ import (
 // See CookieStore and FilesystemStore for examples.
 type Store interface {
 	// Get should return a cached session.
-	Get(r *http.Request, name string) (*Session, error)
+	Get(r *http.Request, name string) (*Session, *http.Request, error)
```
Comment (elithrar) [inline store.go]: Ditto. Breaking the `Store` interface is a major break - it requires third-party store to update, which may be out of the control of the package user.

**Comment by elithrar** [comment]
Further: we should put a note in the README that indicates *why* this is breaking, and the steps to update your code.

**Comment by shawnps** [comment]
> Store changes are not. I don't believe they are required.

@elithrar the tests fail without the changes in the second commit, unless I'm misunderstanding something:

```
➜  sessions [go17-context] git rev-parse HEAD
e1ef5d16938bbb62d2404dd30061606c3725ef7e
➜  sessions [go17-context] go test
PASS
ok  	github.com/gorilla/sessions	0.014s
➜  sessions [go17-context] git checkout ecfa348961f780c90ace75df52f8c588f3762aef
➜  sessions  go test
--- FAIL: TestFlashes (0.00s)
	sessions_test.go:71: No cookies. Header: map[]
FAIL
exit status 1
FAIL	github.com/gorilla/sessions	0.014s
```

**Comment by elithrar** [comment]
The internal changes are required, but you shouldn't need to change the method signature. You already have a request, so no need to return one.

**Comment by shawnps** [comment]
Sorry, I'm not sure what to do from here. The code didn't work until I made the Get methods return the request, so If I revert the signature of `Store.Get` then I have to revert all of the implementations of it, which makes the test fail.

**Comment by shawnps** [comment]
@elithrar example:

```
diff --git a/sessions_test.go b/sessions_test.go
index f00813e..c166b05 100644
--- a/sessions_test.go
+++ b/sessions_test.go
@@ -48,7 +48,7 @@ func TestFlashes(t *testing.T) {
 	req, _ = http.NewRequest("GET", "http://localhost:8080/", nil)
 	rsp = NewRecorder()
 	// Get a session.
-	if session, req, err = store.Get(req, "session-key"); err != nil {
+	if session, err = store.Get(req, "session-key"); err != nil {
 		t.Fatalf("Error getting session: %v", err)
 	}
 	// Get a flash.
@@ -71,7 +71,7 @@ func TestFlashes(t *testing.T) {
 		t.Fatal("No cookies. Header:", hdr)
 	}
 
-	if _, req, err = store.Get(req, "session:key"); err.Error() != "sessions: invalid character in cookie name: session:key" {
+	if _, err = store.Get(req, "session:key"); err.Error() != "sessions: invalid character in cookie name: session:key" {
 		t.Fatalf("Expected error due to invalid cookie name")
 	}
 
@@ -81,7 +81,7 @@ func TestFlashes(t *testing.T) {
 	req.Header.Add("Cookie", cookies[0])
 	rsp = NewRecorder()
 	// Get a session.
-	if session, req, err = store.Get(req, "session-key"); err != nil {
+	if session, err = store.Get(req, "session-key"); err != nil {
 		t.Fatalf("Error getting session: %v", err)
 	}
 	// Check all saved values.
@@ -114,7 +114,7 @@ func TestFlashes(t *testing.T) {
 	req, _ = http.NewRequest("GET", "http://localhost:8080/", nil)
 	rsp = NewRecorder()
 	// Get a session.
-	if session, req, err = store.Get(req, "session-key"); err != nil {
+	if session, err = store.Get(req, "session-key"); err != nil {
 		t.Fatalf("Error getting session: %v", err)
 	}
 	// Get a flash.
@@ -141,7 +141,7 @@ func TestFlashes(t *testing.T) {
 	req.Header.Add("Cookie", cookies[0])
 	rsp = NewRecorder()
 	// Get a session.
-	if session, req, err = store.Get(req, "session-key"); err != nil {
+	if session, err = store.Get(req, "session-key"); err != nil {
 		t.Fatalf("Error getting session: %v", err)
 	}
 	// Check all saved values.
diff --git a/store.go b/store.go
index f041a6b..6c901d6 100644
--- a/store.go
+++ b/store.go
@@ -21,7 +21,7 @@ import (
 // See CookieStore and FilesystemStore for examples.
 type Store interface {
 	// Get should return a cached session.
-	Get(r *http.Request, name string) (*Session, *http.Request, error)
+	Get(r *http.Request, name string) (*Session, error)
 
 	// New should create and return a new session.
 	//
@@ -76,12 +76,12 @@ type CookieStore struct {
 //
 // It returns a new session and an error if the session exists but could
 // not be decoded.
-func (s *CookieStore) Get(r *http.Request, name string) (*Session, *http.Request, error) {
+func (s *CookieStore) Get(r *http.Request, name string) (*Session, error) {
 	var reg *Registry
 	reg, r = GetRegistry(r)
 	ses, err := reg.Get(s, name)
 
-	return ses, r, err
+	return ses, err
 }
 
 // New returns a session for the given name without adding it to the registry.
@@ -183,12 +183,12 @@ func (s *FilesystemStore) MaxLength(l int) {
 // Get returns a session for the given name after adding it to the registry.
 //
 // See CookieStore.Get().
-func (s *FilesystemStore) Get(r *http.Request, name string) (*Session, *http.Request, error) {
+func (s *FilesystemStore) Get(r *http.Request, name string) (*Session, error) {
 	var reg *Registry
 	reg, r = GetRegistry(r)
 	ses, err := reg.Get(s, name)
 
-	return ses, r, err
+	return ses, err
 }
 
 // New returns a session for the given name without adding it to the registry.
```


```
➜  sessions [go17-context] go test
--- FAIL: TestFlashes (0.00s)
	sessions_test.go:71: No cookies. Header: map[]
FAIL
exit status 1
FAIL	github.com/gorilla/sessions	0.014s
```

**Comment by shawnps** [comment]
Btw, I also tried setting `reg.request = r` in the `Get` method but that did not work.

**Comment by elithrar** [comment]
https://github.com/gorilla/sessions/tree/pr95 @shawnps

**Comment by shawnps** [comment]
@elithrar got it, I'll push an update

**Comment by shawnps** [comment]
@elithrar does the new stuff look okay? Anything I can do?

**Comment by elithrar** [comment]
Thinking it through. I'd like to automate some PR creation for a few people
this would break and/or add some notes to the README to explain the
breakage more.

**Comment by shawnps** [comment]
@elithrar ah you did mention the README. I'll add something to this PR updating the README, so please take a look when you have time!

As for the PR creation for people this will break, I can investigate that as well if you'd like. I have a script that automatically makes PRs after running `gofmt -s -w` on a repo, so I could use a modified version of that which takes a list of repos and automatically makes a PR updating `GetRegistry` calls. I guess we could use GitHub's code search for people calling `GetRegistry` to get the list of repos.

**Comment by shawnps** [comment]
Might be able to use something like this to automate, or just use `sed`:

https://github.com/golang/tools/blob/master/refactor/rename/rename.go#L38

**Comment by shawnps** [comment]
@elithrar I just added a note to the README, see the rendered Markdown here:

https://github.com/shawnps/sessions/blob/dc17d8cc4b7165432d37bd224ae43309f97950a9/README.md#notes

**Comment by shawnps** [comment]
(comments/edit suggestions welcome of course)

**Comment by nitingupta910** [comment]
Following this long conversation, it appears changes are now ready to be pulled?

**Comment by elithrar** [comment]
We need to:

a) tag the current HEAD before we break
b) break it
c) tag the new HEAD post break

We also need to document it and reach out to those using the library
without vendoring (the majority of mux users).

**Comment by omeid** [comment]
Just a friendly ping. Is there anything needs to be done to merge this in?

**Comment by elithrar** [comment]
@omeid I'd like to do more work to automatically create PRs on those who import us, don't vendor, and thus will have their applications break. 

gorilla/sessions pre-dates most vendoring tools and I suspect the intersection of those who use it and don't vendor is therefore very high.

**Comment by elithrar** [comment]
Take note of https://github.com/gorilla/sessions/issues/105 - we may just fix it there.

File: sessions.go [hunk 4]
```diff
@@ -186,7 +184,9 @@ func init() {
 
 // Save saves all sessions used during the current request.
 func Save(r *http.Request, w http.ResponseWriter) error {
-	return GetRegistry(r).Save(w)
+	var reg *Registry
+	reg, r = GetRegistry(r)
```
Comment (seh) [inline sessions.go]: If there was no registry saved in the request's context, we shouldn't create a new, empty one here, only to throw it away after telling it to save nothing. Instead, separate `GetRegistry` into its two cases—one that returns an existing registry if it already exists, and one that wraps it and installs a fresh registry if there wasn't one before. This function should use the former and bail early if there was no existing registry.

**Pull request closed without merging**